datacollectiondemo/source/cpp/src/KaaDemo.cpp

```diff
@@ -14,77 +14,111 @@
+            std::make_shared<RecordCountLogUploadStrategy>(int(logUploadThreshold), kaaClient_->getKaaClientContext()));
+
+        // Set up a configuration subsystem.
+        IConfigurationStoragePtr storage(
+            std::make_shared<FileConfigurationStorage>(std::string(savedConfig_)));
+        kaaClient_->setConfigurationStorage(storage);
+        kaaClient_->addConfigurationListener(*this);
+    }
 
-        futurePairs.push_back(std::make_pair(std::move(kaaClient->addLogRecord(logRecord)), logRecord.timeStamp));
-        std::cout << "Sent " << logNumber << "th record" << std::endl;
+    ~TemperatureSensor()
+    {
+        // Stop the Kaa endpoint.
+        kaaClient_->stop();
+        std::lock_guard<std::mutex> guard(iostream_mutex_);
+        std::cout << "Data collection demo stopped" << std::endl;
     }
 
-    for (auto& pair : futurePairs) {
-        try {
-            RecordInfo recordInfo = pair.first.get();
-            BucketInfo bucketInfo = recordInfo.getBucketInfo();
+    void run()
+    {
+        // Run the Kaa endpoint.
+        kaaClient_->start();
+        samplePeriod_ = kaaClient_->getConfiguration().samplePeriod;
+        timer_.async_wait(boost::bind(&TemperatureSensor::sendTemperature, this));
+        timer_.expires_from_now(boost::posix_time::seconds(samplePeriod_));
+        service_.run();
+    }
 
-            std::size_t timeSpent = (recordInfo.getRecordAddedTimestampMs() - pair.second)
-                                        + recordInfo.getRecordDeliveryTimeMs();
+private:
+    static constexpr auto savedConfig_ = "saved_config.cfg";
+    static constexpr int logUploadThreshold = 5;
+    std::shared_ptr<IKaaClient> kaaClient_;
+    int32_t samplePeriod_;
+    boost::asio::io_service service_;
+    boost::posix_time::seconds interval_;
+    boost::asio::deadline_timer timer_;
+    std::mutex iostream_mutex_;
+
+    int32_t getTemperature()
+    {
+        // For sake of example random data is used
+        return rand() % 10 + 25;
+    }
 
-            std::cout << "Received log record delivery info. Bucket Id [" <<  bucketInfo.getBucketId() << "]. "
-                      << "Record delivery time [" << timeSpent << " ms]." << std::endl;
-        } catch (std::exception& e) {
-            std::cout << "Exception was caught while waiting for callback future" << e.what();
+    void sendTemperature()
+    {
+        KaaUserLogRecord logRecord;
+        logRecord.temperature = getTemperature();
+        logRecord.timeStamp = std::time(nullptr);
+        // Send value of temperature
+        kaaClient_->addLogRecord(logRecord);
+        // Show log
+        {
+            std::lock_guard<std::mutex> guard(iostream_mutex_);
+            std::cout << "Sampled temperature: " << logRecord.temperature
+                << ", timestamp: " << logRecord.timeStamp << std::endl;
         }
+        // Change timer expiry period
+        timer_.expires_from_now(boost::posix_time::seconds(samplePeriod_));
```

Comment:
In this case you have a delay, the period will not be quite accurate. You can avoid it this way:
` timer_->expires_at(timer_->expires_at() + boost::posix_time::seconds(samplePeriod_));